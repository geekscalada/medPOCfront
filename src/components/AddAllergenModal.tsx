import React, { useEffect, useState } from "react";
import { IonButton, IonSearchbar, IonItem, IonList } from "@ionic/react";
import { CustomModalComponent } from "../services/customModalComposer";
import { AxiosRequestConfig } from "axios";
import useApiDebouncedRequest from "../services/useApiDebouncedRequest";
import { ArrayAllergens } from "../models/types/types";
import ConfirmationAllergenSheeAction from "./ConfirmationAllergernModal";

import { ModalComposer } from "../services/customModalComposer";
import useModalHelper from "../hooks/useModalHelper";

import useGlobalInfoContext from "../hooks/useGlobalInfoContext";

export interface AddAllergenComponentModalProps {
  isOpen: boolean;
  onClose: () => void;
}

const AddAllergenComponentModal: React.FC<AddAllergenComponentModalProps> = ({
  isOpen,
  onClose,
}) => {
  const [searchTerm, setSearchTerm] = useState("");
  const { isModalOpen, openModal, closeModal } = useModalHelper();
  const apiUrl = import.meta.env.VITE_API_URL;
  const [componentes, setComponentes] = useState<string[]>([]);

  const { headerTitle, setHeaderTitle } = useGlobalInfoContext();

  /**
   * Logic to confirmation modal
   */
  const [isConfirmationModalOpen, setIsConfirmationModalOpen] = useState(false);
  const [selectedAllergen, setSelectedAllergen] = useState("");
  const closeConfirmationModal = () => {
    setIsConfirmationModalOpen(false);
  };

  const optionsGet: AxiosRequestConfig = {
    method: "GET",
    url: apiUrl + "/alergenos/" + searchTerm,
    headers: { "Content-Type": "application/json" },
  };

  // This hook refreshes the data when the searchTerm changes
  // TODO: Implement loadings
  const { data, loading, error } = useApiDebouncedRequest<ArrayAllergens>(
    optionsGet,
    2
  );

  useEffect(() => {
    // Cambiar miVariable justo después de la primera renderización
    setHeaderTitle("Add Allergen");
  }, []);

  useEffect(() => {
    if (isOpen) {
      openModal();
    } else {
      closeModal();
    }
  }, [isOpen]);

  useEffect(() => {
    // We refresh data when hook of useApiDebouncedRequest changes

    if (error) {
      console.log("Error");
      return;
    }

    setComponentes(data?.allergens ?? []);
  }, [data, error]);

  const modalComposer: ModalComposer = {
    titleModal: "Añadir alérgeno",
    headerModal: null,
    content: (
      <>
        <div className="ion-padding">
          <IonItem>
            <IonSearchbar
              color={"medium"}
              debounce={500}
              data-testid="allergens-searchbar"
              placeholder="Search"
              onIonInput={(e) => setSearchTerm(e.detail.value!)}
              value={searchTerm}
              onIonClear={() => setSearchTerm("")}
            ></IonSearchbar>
          </IonItem>
          <IonList>
            {componentes.map((componente) => (
              <IonItem
                key={componente + "id"}
                button
                onClick={() => {
                  setSelectedAllergen(componente);
                  setIsConfirmationModalOpen(true);
                }}
              >
                {componente} {/* Ajusta según la estructura de tus datos */}
              </IonItem>
            ))}
          </IonList>
        </div>
      </>
    ),
    buttons: [],
    modalSize: {
      width: "550px",
      height: "650px",
    },
    canDismissModal: true,
    closeButtonHeader: true,
  };

  return (
    <>
      <CustomModalComponent
        modalComposer={modalComposer}
        isOpen={isModalOpen}
        onClose={onClose} // Callback
      />
      <ConfirmationAllergenSheeAction
        isOpen={isConfirmationModalOpen}
        allergen={selectedAllergen}
        onClose={closeConfirmationModal}
      />
    </>
  );
};

export default AddAllergenComponentModal;
