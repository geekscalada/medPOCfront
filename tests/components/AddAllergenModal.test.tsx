import { describe, it, expect, vi } from "vitest";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import AddAllergenModal from "../../src/components/AddAllergenModal";
import React from "react";

describe("AddAllergenModal Component", () => {
  // Función simulada para cerrar el modal. Viene de Vitest y se usa para simular funciones y su llamada.
  const mockCloseModal = vi.fn();

  it("is rendered correctly", async () => {
    // 'render' renderiza el componente en un entorno virtual para poder interactuar con él en las pruebas.
    // 'asFragment' toma una instantánea del DOM renderizado por el componente. Esto es útil para comparar cambios en el DOM a lo largo del tiempo.
    const { asFragment } = render(
      <AddAllergenModal closeModal={mockCloseModal} />
    );

    // 'expect' con 'toMatchSnapshot' comprueba que el fragmento capturado coincida con una versión anterior guardada. Si no hay una versión anterior, guarda la actual.
    expect(asFragment()).toMatchSnapshot();
  });

  it("update the searchterm when we write on the searchbar", async () => {
    render(<AddAllergenModal closeModal={mockCloseModal} />);

    // 'screen' es una utilidad de '@testing-library/react' que proporciona métodos para seleccionar elementos del DOM renderizado, similar a cómo lo harías en un navegador.
    const searchBar = screen.getByTestId("allergens-searchbar");

    // 'fireEvent' es una utilidad que simula eventos del usuario en la interfaz. En este caso, simula la escritura en un campo de entrada (input).
    // Tiene click, etc.
    fireEvent.change(searchBar, { target: { value: "Milk" } });

    // 'waitFor' se utiliza para esperar a que aparezca un elemento en el documento, útil para pruebas que involucran actualizaciones asíncronas del DOM.
    // ToBeInTheDocument es una utilidad de '@testing-library/jest-dom' que verifica si un elemento está presente en el documento.
    // getByDisplayValue es una utilidad de '@testing-library/react' que selecciona un elemento por su valor.
    await waitFor(() => {
      expect(screen.getByDisplayValue("Milk")).toBeInTheDocument();
    });
  });
});

//TODO:
// Additional tests could include:

// - Verifying the behavior when the search term is cleared.
// - Testing error handling if the data fetching fails.
// - Ensuring the add button behaves correctly.
