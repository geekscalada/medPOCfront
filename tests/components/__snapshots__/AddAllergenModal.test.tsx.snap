// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddAllergenModal Component > is rendered correctly 1`] = `
<DocumentFragment>
  <ion-content>
    <ion-header>
      <!---->
      <ion-toolbar>
        <ion-title
          class="ion-text-center"
        >
          Añadir Alérgeno
        </ion-title>
      </ion-toolbar>
    </ion-header>
    <div
      class="ion-padding"
    >
      <ion-item>
        <ion-searchbar
          data-testid="allergens-searchbar"
          debounce="500"
          placeholder="Busca un alérgeno"
          value=""
        />
      </ion-item>
      <ion-list />
      <ion-button
        class="ion-margin-top"
        expand="block"
      >
        Añadir
      </ion-button>
      <ion-button
        class="ion-margin-top"
        color="medium"
        expand="block"
      >
        Cancelar
      </ion-button>
    </div>
    <template />
  </ion-content>
</DocumentFragment>
`;
